BPOR/BPOR.Rms/Services/CampaignService.cs

```diff
@@ -216,57 +218,99 @@ private void UpdateProcessingResultReferences(ProcessingResults processingResult
                 continue;
             }
 
-            var reference = emailQueue.SelectMany(e => e.StudyParticipantEnrollments)
+            var reference = queue.SelectMany(e => e.StudyParticipantEnrollments)
                 .FirstOrDefault(e => e.ParticipantId == volunteer.Id)?.Reference;
 
             if (reference == null)
             {
                 logger.LogWarning("Reference not found for volunteer {VolunteerId}", volunteer.Id);
-                reference = $"no-study.{emailCampaignParticipant.Id}";
+                reference = $"no-study.{campaignParticipant.Id}";
             }
 
             var baseUri = new Uri(callback);
             var queryParams = new Dictionary<string, string>
             {
-                { "reference", encryptionService.Encrypt(emailCampaignParticipant.Id.ToString()) }
+                { "reference", encryptionService.Encrypt(campaignParticipant.Id.ToString()) }
             };
 
             var uriWithQuery = new Uri(QueryHelpers.AddQueryString(baseUri.ToString(), queryParams));
             var link = uriWithQuery.ToString();
             var notification = new Notification
             {
-                PrimaryIdentifier = volunteer.Email,
                 NotificationDatas = new List<NotificationData>
                 {
-                    new() { Key = "email", Value = volunteer.Email },
-                    new() { Key = "emailCampaignParticipantId", Value = emailCampaignParticipant.Id.ToString() },
+                    new() { Key = "campaignParticipantId", Value = campaignParticipant.Id.ToString() },
+                    new() { Key = "campaignTypeId", Value = campaignParticipant.CampaignTypeId.ToString() },
                     new() { Key = "firstName", Value = volunteer.FirstName },
                     new() { Key = "lastName", Value = volunteer.LastName },
-                    new()
-                    {
-                        Key = "uniqueLink",
-                        Value = link
-                    },
-                    new() { Key = "emailTemplateId", Value = campaign.EmailTemplateId.ToString() },
+                    new() { Key = "uniqueLink", Value = link },
+                    new() { Key = "templateId", Value = campaign.TemplateId.ToString() },
                     new() { Key = "uniqueReference", Value = reference }
                 }
             };
 
+            switch (campaign.TypeId)
+            {
+                case (int)ContactMethod.Email:
+                    notification.PrimaryIdentifier = volunteer.Email;
+                    notification.NotificationDatas.Add(new NotificationData { Key = "email", Value = volunteer.Email });
+                    break;
+
+                case (int)ContactMethod.Letter:
+                    if (string.IsNullOrWhiteSpace(volunteer.Address.AddressLine1) ||
+                        string.IsNullOrWhiteSpace(volunteer.Address.AddressLine2) ||
```

Comment:
similar to other comment, address line 2 is a optional field for VS and RMS sign up.

Reply:
the requirement is 3 address lines. VS currently requires 3 (address line 1, town, postcode) which is fine with Notify.